core/src/main/java/jeeves/transaction/TransactionManager.java

```diff
@@ -105,65 +168,255 @@
+     * @param status transaction status
+     * @throws Throwable
+     */
+    private static void fireBeforeCommit(ApplicationContext context,
+                                         PlatformTransactionManager transactionManager,
+                                         @Nullable TransactionStatus status) throws Throwable {
+        Throwable beforeCommitFailure = null;
 
-        for (AfterCommitTransactionListener listener : context.getBeansOfType(AfterCommitTransactionListener.class).values()) {
-            listener.afterCommit(transaction);
+        Collection<BeforeCommitTransactionListener> listeners = context.getBeansOfType
+            (BeforeCommitTransactionListener.class).values();
+        for (BeforeCommitTransactionListener listener : listeners) {
+            try {
+                listener.beforeCommit(status);
+            }
+            catch (Throwable t){
+                Log.debug(Log.JEEVES, "Listener "+listener.toString()+" newTransaction callback failed: "+ t);
+                beforeCommitFailure = t;
+            }
+        }
+        if (beforeCommitFailure != null){
+            throw beforeCommitFailure;
         }
     }
 
+    /**
+     * Notify context NewTransactionListener instances of transaction status change.
+     *
+     * @param context context used to obtain listeners
+     * @param transactionManager manager responsible for execution
+     * @param status transaction status
+     * @throws Throwable
+     */
+    private static void fireNewTransaction(ApplicationContext context,
+                                           PlatformTransactionManager transactionManager,
+                                           @Nullable TransactionStatus status) throws Throwable {
+        Throwable newTransactionFailure = null;
+
+        Collection<NewTransactionListener> listeners = context.getBeansOfType
+            (NewTransactionListener.class).values();
+        for (NewTransactionListener listener : listeners) {
+            try {
+                listener.newTransaction(status);
+            }
+            catch (Throwable t){
+                Log.debug(Log.JEEVES, "Listener "+listener.toString()+" newTransaction callback failed: "+ t);
+                newTransactionFailure = t;
+            }
+        }
+        if (newTransactionFailure != null){
+            throw newTransactionFailure;
+        }
+    }
+
+    /**
+     * Notify context AfterRollbackTransactionListener instances of transaction status change.
+     *
+     * @param context context used to obtain listeners
+     * @param transactionManager manager responsible for execution
+     * @param status transaction status
+     * @throws Throwable
+     */
+    private static void fireAfterRollback(ApplicationContext context,
+                                          PlatformTransactionManager transactionManager,
+                                          @Nullable TransactionStatus status) throws Throwable {
```

Comment:
While I could of made use of generics (passing in the class of transaction listener) having a clear method for each notification in having a clear spot to debug.